CRC Generation Module

ronitnag04 opened the pull request.

ronitnag04 committed:
Wrote initial module structure

ronitnag04 committed:
Added Reset Logic

ronitnag04 committed:
Generated CRC-16 Lookup table

ronitnag04 committed:
Created lookup table class for CRC calculations

ronitnag04 committed:
Fixed lookup table format and successful import

ronitnag04 committed:
Set up CRC Gen algorithm (not functional)

ronitnag04 committed:
Updated CRCGenerator for Shift Register (non-functional)

ronitnag04 committed:
Compiled, functioning shift, algorithm incorrect

ronitnag04 committed:
Test Structure fixed

ronitnag04 committed:
Resolved issues, added test cases

ronitnag04 committed:
Added test cases

ronitnag04 committed:
style: Refactored to snake_case, split test modules

ronitnag04 committed:
Refactored data_in to message Decoupled ReadyValid3IO

ronitnag04 committed:
Remove helper files

ronitnag04 committed:
Format changes

ronitnag04 committed:
Format with scalafmtAll

ronitnag04 committed:
Added back-to-back 1024-bit test

ronitnag04 committed:
Reconfigured CRC 3IO interface, documentation

ronitnag04 committed:
Fixed documentation typo

ansaschmulbach commented on src/main/scala/d2dadapter/CRCGenerator.scala:
```diff
@@ -0,0 +1,132 @@
+  val crc1 = RegInit(Bits(8.W), 0.U)
+
+  // Output signal registers
+  val message_ready = RegInit(Bool(), true.B)
```
nit: camelcase?

ronitnag04 replied:
I think I was told to use snake case by @rahulk29, but I can switch it to camelcase if that is recommended.

ansaschmulbach commented on src/main/scala/d2dadapter/CRCGenerator.scala:
```diff
@@ -0,0 +1,132 @@
+      * and crc valid state low. CRC bits will be cleared to 0.
+      * @group Signals
+      */
+    val reset = Input(Bool())
```
use implicit reset instead

ronitnag04 replied:
Would I just be able to omit the reset declaration in the IO. I also changed it so that when a new message is latched, the crc output is cleared and set invalid, let me know if this is correct though.

ansaschmulbach commented on src/main/scala/d2dadapter/CRCGenerator.scala:
```diff
@@ -0,0 +1,132 @@
+
+  /* Idle with crc output valid until consumer initiates transfer of CRC with
+   * io.crc.ready */
+  when(crc_valid) {
```
why are we not using fire here instead?

ethanwu10 reviewed (commented):
In terms of architecture, we will probably need something that can run faster than one byte per cycle in order to keep up with the link's speed, and also it might be good to eventually have some sort of streaming capability so that we can feed the module bytes as soon as they are available, instead of collecting the entire word and summing it at once. This is fine for now though and we should probably wait until we better understand how the module will get used to determine what architecture changes to make.

ethanwu10 commented on src/main/scala/d2dadapter/CRCGenerator.scala:
```diff
@@ -0,0 +1,132 @@
+      */
+    val reset = Input(Bool())
+
+    /** ReadyValid3IO interface to allow consumer to transfer message bits to
```
This is a normal `ReadValid` interface, not `ReadyValid3` / `Decoupled3` (that is just for UCIe's special version ready/valid with the `irdy` signal in addition to normal `ready` and `valid`)

(same thing for other references to `ReadyValid3IO`

ronitnag04 committed:
Reconfigured IO, testing with DecoupledDriver

ronitnag04 committed:
Fixed lookup table docstring

ronitnag04 commented on src/main/scala/d2dadapter/CRCGenerator.scala:
```diff
@@ -0,0 +1,88 @@
+
+  // Computation not finished when step is not 0
+  when(step > 0.U) {
+    crc1 := crc0 ^ lookup
+      .table(crc1 ^ message_bits(width - 1, width - 8))(15, 8)
+    crc0 := lookup.table(crc1 ^ message_bits(width - 1, width - 8))(7, 0)
+    message_bits := message_bits << 8
```
Yeah, it's a little unconventional compared to what I normally see for CRC calculations, but the results of this CRC generator match the results of the design the UCIe1.0 spec has in the appendix (where they just selectively xor the 1024 bits of the message for each of the 16 crc bits). I ran a simulation of the xor'ing method in python to generate the test cases I listed in the test module.

ronitnag04 commented on src/main/scala/d2dadapter/CRC16Lookup8Bit.scala:
```diff
@@ -0,0 +1,267 @@
+
+import chisel3._
+
+class CRC16Lookup {
```
I tried switching it to an object but I am running into issues with trying to access the data from the CRCGenerator module (it says the wire is not accessible from the CRCGenerator module). I'll try to look more into the documentation to find the right way to do this, but if you have an idea of the issue I am running into I would appreciate some guidance.

ethanwu10 replied:
I've fixed this for you by making the field a method instead, so that the `VecInit` is created at the use site inside the module context (instead of when the singleton object is constructed)

ronitnag04 committed:
Fixed with scalafmtAll

ethanwu10 committed:
refactor: make CRC16Lookup an object

Since hardware types can only be instantiated inside of a module, the
`table` member is made into a parentheses-less method so that the
`VecInit` is constructed at the use site.

ronitnag04 committed:
Fixed error test case

ronitnag04 committed:
Reconfigured CRC update to one line

ronitnag04 committed:
Added paramterized bytes consumed per cycle

ronitnag04 committed:
Merge branch 'main' into d2dadapter-crc

ronitnag04 committed:
Refactored to only send next-consumed message bits

ronitnag04 committed:
Fixed formatting

ronitnag04 committed:
Replaced CRC lookup table with shift, XOR CRC_poly method

ronitnag04 committed:
Merge branch 'main' into d2dadapter-crc

ronitnag04 committed:
Merge branch 'd2dadapter-crc' of https://github.com/ucb-ucie/uciedigital into d2dadapter-crc
